server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java

```diff
@@ -67,37 +58,93 @@ public long encode(double x, double y, int precision) {
-            long numLatCells = Math.max(1, (long) Math.ceil(
-                (bounds.maxY() - geoHashCell.getMinY()) / Geohash.latHeightInDegrees(precision)));
-            return numLonCells * numLatCells;
+            assert bounds.minX() <= bounds.maxX();
+            long numLonCells = (long) ((bounds.maxX() - bounds.minX()) / Geohash.lonWidthInDegrees(precision));
+            long numLatCells = (long) ((bounds.maxY() - bounds.minY()) / Geohash.latHeightInDegrees(precision));
+            long count = (numLonCells + 1) * (numLatCells + 1);
+            if (count == 1) {
+                String hash = Geohash.stringEncode(bounds.minX(), bounds.minY(), precision);
+                values.resizeCell(1);
+                values.add(0, Geohash.longEncode(hash));
+                return 1;
+            } else if (count <= precision) {
+                return setValuesByBruteForceScan(values, geoValue, precision, bounds);
+            } else {
+                return setValuesByRasterization("", values, 0, precision, geoValue);
+            }
         }
 
-        @Override
-        public int setValues(long[] values, MultiGeoValues.GeoValue geoValue, int precision) {
-            MultiGeoValues.BoundingBox bounds = geoValue.boundingBox();
+        protected int setValuesByBruteForceScan(CellIdSource.CellValues values, MultiGeoValues.GeoValue geoValue,
+                                                int precision, MultiGeoValues.BoundingBox bounds) {
             int idx = 0;
-            // find minimum (x,y) of geo-hash-cell that contains (bounds.minX, bounds.minY)
-            String hash = Geohash.stringEncode(bounds.minX(), bounds.minY(), precision);
-            Rectangle geoHashCell = Geohash.toBoundingBox(hash);
-            for (double i = geoHashCell.getMinX(); i < bounds.maxX(); i+= Geohash.lonWidthInDegrees(precision)) {
-                for (double j = geoHashCell.getMinY(); j < bounds.maxY(); j += Geohash.latHeightInDegrees(precision)) {
+            String min = Geohash.stringEncode(bounds.minX(), bounds.minY(), precision);
+            String max = Geohash.stringEncode(bounds.maxX(), bounds.maxY(), precision);
+            double minY = Geohash.decodeLatitude(min);
+            double minX = Geohash.decodeLongitude(min);
+            double maxY = Geohash.decodeLatitude(max);
+            double maxX = Geohash.decodeLongitude(max);
+            for (double i = minX; i <= maxX; i += Geohash.lonWidthInDegrees(precision)) {
+                for (double j = minY; j <= maxY; j += Geohash.latHeightInDegrees(precision)) {
                     Rectangle rectangle = Geohash.toBoundingBox(Geohash.stringEncode(i, j, precision));
                     GeoRelation relation = geoValue.relate(rectangle);
                     if (relation != GeoRelation.QUERY_DISJOINT) {
-                        values[idx++] = encode(i, j, precision);
+                        values.resizeCell(idx + 1);
```

Comment:
should we be more coarse here? The way this code works, wouldn't we re-allocate and copy the `long[]` potentially 10,000 times?

Reply:
Not really, the way the arrays are re-allocated is not lineal:

```
  /** 
   * Returns an array size >= minTargetSize, generally
   *  over-allocating exponentially to achieve amortized
   *  linear-time cost as the array grows.
   */
```